packages/flutter_tools/lib/src/runner/local_engine.dart

```diff
@@ -0,0 +1,154 @@
+import 'package:package_config/package_config.dart';
+
+import '../artifacts.dart';
+import '../base/common.dart';
+import '../base/file_system.dart';
+import '../base/logger.dart';
+import '../base/platform.dart';
+import '../base/user_messages.dart' hide userMessages;
+import '../dart/package_map.dart';
+import 'flutter_command_runner.dart';
+
+/// A strategy for locating the out/ directory of a local engine build.
+///
+/// The flutter tool can be run with the output files of one or more engine builds
+/// replacing the cached artifacts. Typically this is done by setting the
+/// `--local-engine` command line flag to the name of the desired engine variant
+/// (e.g. "host_debug_unopt").  Provided that the `flutter/` and `engine/` directories
+/// are located adjacent to one another, the output folder will be located
+/// automatically.
+///
+/// For scenarios where the engine is not adjacent to flutter, the
+/// `--local-engine-src-path` can be provided to give an exact path.
+///
+/// For more information on local engines, see CONTRIBUTING.md.
+class LocalEngineLocator {
+  LocalEngineLocator({
+    @required Platform platform,
+    @required Logger logger,
+    @required FileSystem fileSystem,
+    @required String flutterRoot,
+    @required UserMessages userMessages,
+  }) : _platform = platform,
+       _logger = logger,
+       _fileSystem = fileSystem,
+       _flutterRoot = flutterRoot,
+        _userMessages = userMessages;
+
+  final Platform _platform;
+  final Logger _logger;
+  final FileSystem _fileSystem;
+  final String _flutterRoot;
+  final UserMessages _userMessages;
+
+  /// Returns the engine build path of a local engine if one is located, otherwise `null`.
+  Future<EngineBuildPaths> findEnginePath(String engineSourcePath, String localEngine) async {
+    engineSourcePath ??= _platform.environment[kFlutterEngineEnvironmentVariableName];
+
+    if (engineSourcePath == null && localEngine != null) {
+      try {
+        final PackageConfig packageConfig = await loadPackageConfigWithLogging(
+          _fileSystem.file(globalPackagesPath),
+          logger: _logger,
+          throwOnError: false,
+        );
+        // Skip if sky_engine is the version in bin/cache.
+        Uri engineUri = packageConfig[kFlutterEnginePackageName]?.packageUriRoot;
+        final String cachedPath = _fileSystem.path.join(_flutterRoot, 'bin', 'cache', 'pkg', kFlutterEnginePackageName, 'lib');
+        if (engineUri != null && _fileSystem.identicalSync(cachedPath, engineUri.path)) {
+          engineUri = null;
+        }
+        // If sky_engine is specified and the engineSourcePath not set, try to
+        // determine the engineSourcePath by sky_engine setting. A typical engine Uri
+        // looks like:
+        // file://flutter-engine-local-path/src/out/host_debug_unopt/gen/dart-pkg/sky_engine/lib/
+        if (engineUri?.path != null) {
+          engineSourcePath = _fileSystem.directory(engineUri.path)
+            ?.parent
```

Comment:
nit: I would have expected the question mark to come at the end of the last line rather than the beginning of this one, but if this style is consistent with other uses it should be fine.

Reply:
`?.` is one operator though, splitting them up seems odd - or do you mean leaving both on a previous line?

Reply:
Hm. Dartfmt does indeed move the `?` part to the next line as well. Maybe I'm just starting to think in terms of null-safety world, with `?` as one operator and `.` as another, but this is fine.